br/pkg/restore/client.go

```diff
@@ -1756,26 +1772,37 @@ func (rc *Client) ReadStreamMetaByTS(ctx context.Context, shiftedStartTS uint64,
 func (rc *Client) ReadStreamDataFiles(
 	ctx context.Context,
 	metas []*backuppb.Metadata,
-) (dataFile, metaFile []*backuppb.DataFileInfo, err error) {
+) (dataFiles, metaFiles []*backuppb.DataFileInfo, err error) {
 	dFiles := make([]*backuppb.DataFileInfo, 0)
 	mFiles := make([]*backuppb.DataFileInfo, 0)
 
 	for _, m := range metas {
-		for _, d := range m.Files {
-			if d.MinTs > rc.restoreTS {
-				continue
-			} else if d.Cf == stream.WriteCF && d.MaxTs < rc.startTS {
-				continue
-			} else if d.Cf == stream.DefaultCF && d.MaxTs < rc.shiftStartTS {
-				continue
-			}
+		for _, ds := range m.FileGroups {
+			for _, d := range ds.DataFilesInfo {
+				if d.MinTs > rc.restoreTS {
+					continue
+				} else if d.Cf == stream.WriteCF && d.MaxTs < rc.startTS {
+					continue
+				} else if d.Cf == stream.DefaultCF && d.MaxTs < rc.shiftStartTS {
+					continue
+				}
 
-			if d.IsMeta {
-				mFiles = append(mFiles, d)
-			} else {
-				dFiles = append(dFiles, d)
+				// If ds.Path is empty, it is MetadataV1.
+				if m.MetaVersion == backuppb.MetaVersion_V2 {
+					d.Path = ds.Path
+				}
+
+				if d.IsMeta {
+					mFiles = append(mFiles, d)
+				} else {
+					dFiles = append(dFiles, d)
+				}
+				log.Debug("backup stream collect data partition", zap.Uint64("offset", d.Offset), zap.Uint64("length", d.Length))
+			}
+			// metadatav1 doesn't use cache
+			if m.MetaVersion == backuppb.MetaVersion_V2 {
+				rc.helper.InitCacheEntry(ds.Path, len(ds.DataFilesInfo))
```

Comment:
If a file group contains not only meta files, can this cache be released?

Reply:
TiKV will seperate ddl and dml files. So in the normal case, it looks good.

But let me make it more robust. :muscle:

Reply:
TiKV will seperate dml and ddl files. In the normal case, it looks ok.
But I find that some files may be filtered out because their xxx-ts is not match.
So in fact, the refcnt should be the number of metaFile appended into `mFiles`.

Reply:
> If a file group contains not only meta files, can this cache be released?

don't worry because BR don't read the dml files.